plugins/ingest/src/test/java/org/elasticsearch/plugin/ingest/transport/IngestActionFilterTests.java

```diff
@@ -225,4 +226,64 @@ public void run() {
         threadPool.shutdown();
     }
 
+    public void testApply_withBulkRequestWithFailure() throws Exception {
+        BulkRequest bulkRequest = new BulkRequest();
+        bulkRequest.putHeader(IngestPlugin.PIPELINE_ID_PARAM, "_id");
+        int numRequest = scaledRandomIntBetween(8, 64);
+        int numNonIndexRequests = 0;
+        for (int i = 0; i < numRequest; i++) {
+            if (i % 2 == 0) {
+                numNonIndexRequests++;
+                ActionRequest request;
+                if (randomBoolean()) {
+                    request = new DeleteRequest("_index", "_type", "_id");
+                } else {
+                    request = new UpdateRequest("_index", "_type", "_id");
+                }
+                bulkRequest.add(request);
+            } else {
+                IndexRequest indexRequest = new IndexRequest("_index", "_type", "_id");
+                indexRequest.source("field1", "value1");
+                bulkRequest.add(indexRequest);
+            }
+        }
+
+        RuntimeException exception = new RuntimeException();
+        Answer answer = (invocationOnMock) -> {
+            PipelineExecutionService.Listener listener = (PipelineExecutionService.Listener) invocationOnMock.getArguments()[2];
+            listener.failed(exception);
+            return null;
+        };
+        doAnswer(answer).when(executionService).execute(any(Data.class), eq("_id"), any(PipelineExecutionService.Listener.class));
+
+        ActionListener actionListener = mock(ActionListener.class);
+        RecordRequestAFC actionFilterChain = new RecordRequestAFC();
+
+        filter.apply("_action", bulkRequest, actionListener, actionFilterChain);
+
+        BulkRequest interceptedRequests = actionFilterChain.getRequest();
+        assertThat(interceptedRequests.requests().size(), equalTo(numNonIndexRequests));
+
+        verifyZeroInteractions(actionListener);
```

Comment:
I'm trying to remove some meat from the `BulkProcessingState` (and yes, this can use a  better name :) ) so that it basically doesn't deal with the chain and listener. That should be handled by the filter itself